imap/mailbox.c

```diff
@@ -3923,6 +3943,233 @@ EXPORTED int mailbox_add_email_alarms(struct mailbox *mailbox)
 }
 #endif // WITH_JMAP
 
+#ifdef USE_SIEVE
+static struct sieve_db *mailbox_open_sieve(struct mailbox *mailbox)
+{
+    if (!mailbox->sievedir) {
+        char *userid = mboxname_to_userid(mailbox_name(mailbox));
+        mailbox->sievedir = xstrdup(user_sieve_path(userid));
+        free(userid);
+    }
+    if (!mailbox->local_sieve) {
+        mailbox->local_sieve = sievedb_open_mailbox(mailbox);
+        int r = sievedb_begin(mailbox->local_sieve);
+        if (r) {
+            sievedb_abort(mailbox->local_sieve);
+            sievedb_close(mailbox->local_sieve);
+            mailbox->local_sieve = NULL;
+        }
+    }
+    return mailbox->local_sieve;
+}
+
+static int mailbox_update_sieve(struct mailbox *mailbox,
+                              const struct index_record *old,
+                              const struct index_record *new)
+{
+    struct sieve_db *sievedb = NULL;
+    struct param *param;
+    struct body *body = NULL;
+    struct sieve_data *sdata = NULL;
+    char *id = NULL;
+    const char *name = NULL;
+    struct buf msg_buf = BUF_INITIALIZER;
+    int isexpunged = (new->internal_flags & FLAG_INTERNAL_EXPUNGED ? 1 : 0);
+    int isactive = (new->system_flags & FLAG_FLAGGED ? 1 : 0);
+    int r = 0;
+
+    if (mbtype_isa(mailbox_mbtype(mailbox)) != MBTYPE_SIEVE) return 0;
+
+    /* never have Sieve on deleted mailboxes */
+    if (mboxname_isdeletedmailbox(mailbox_name(mailbox), NULL)) return 0;
+
+    /* conditions in which there's nothing to do */
+    if (!new) return 0;
+
+    /* phantom record - never really existed here */
+    if (!old && isexpunged) return 0;
+
+    r = mailbox_cacherecord(mailbox, new);
+    if (r) goto done;
+
+    /* Get script id from filename param in Content-Disposition header */
+    message_read_bodystructure(new, &body);
+    for (param = body->disposition_params; param; param = param->next) {
+        if (!strcmp(param->attribute, "FILENAME")) {
+            ssize_t idlen = strlen(param->value) - strlen(SIEVE_EXTENSION);
```

Comment:
If the ID is the filename, do we need to restrict filenames via timsieved to be legit JMAP OBJECTID (length + valid chars)?  Or is this always a separate JMAP-compatible ID?  (I see it's mainly made by uuidgen)

Reply:
The filename param is actually the JMAP ID, so its always a GUID that we generate.